# Word-Level Tokenizing

If not present, download the dataset = "https://raw.githubusercontent.com/karpathy/char-rnn/master/data/tinyshakespeare/input.txt" to `data/01_raw/shakespeare.txt`

In [1]:
%load_ext kedro.ipython
%reload_kedro

from typing import Any, Dict, List, Tuple

import re

from nltk.corpus import stopwords
from sklearn.feature_extraction.text import CountVectorizer

[02/06/24 23:36:46] INFO     Resolved project path as: c:\Users\bcech\Documents\ML\tokenizers.      ]8;id=143550;file://c:\Users\bcech\anaconda3\envs\tok\Lib\site-packages\kedro\ipython\__init__.py\__init__.py]8;;\:]8;id=750724;file://c:\Users\bcech\anaconda3\envs\tok\Lib\site-packages\kedro\ipython\__init__.py#149\149]8;;\
                             To set a different path, run '%reload_kedro <project_root>'                           

[02/06/24 23:36:47] INFO     Kedro project tokenizers                                               ]8;id=434417;file://c:\Users\bcech\anaconda3\envs\tok\Lib\site-packages\kedro\ipython\__init__.py\__init__.py]8;;\:]8;id=603030;file://c:\Users\bcech\anaconda3\envs\tok\Lib\site-packages\kedro\ipython\__init__.py#119\119]8;;\

                    INFO     Defined global variable 'context', 'session', 'catalog' and            ]8;id=877388;file://c:\Users\bcech\anaconda3\envs\tok\Lib\site-packages\kedro\ipython\__init__.py\__init__.py]8;;\:]8;id=670811;file://c:\Users\bcech\anaconda3\envs\tok\Lib\site-packages\kedro\ipython\__init__.py#120\120]8;;\
                             'pipelines'                                                                           

[02/06/24 23:36:49] INFO     Registered line magic 'run_viz'                                        ]8;id=63668;file://c:\Users\bcech\anaconda3\envs\tok\Lib\site-packages\kedro\ipython\__init__.py\__init__.py]8;;\:]8;id=260408;file://c:\Users\bcech\anaconda3\envs\tok\Lib\site-packages\kedro\ipython\__init__.py#126\126]8;;\

                    INFO     Resolved project path as: C:\Users\bcech\Documents\ML\tokenizers.      ]8;id=454707;file://c:\Users\bcech\anaconda3\envs\tok\Lib\site-packages\kedro\ipython\__init__.py\__init__.py]8;;\:]8;id=833555;file://c:\Users\bcech\anaconda3\envs\tok\Lib\site-packages\kedro\ipython\__init__.py#149\149]8;;\
                             To set a different path, run '%reload_kedro <project_root>'                           

                    INFO     Kedro project tokenizers                                               ]8;id=93062;file://c:\Users\bcech\anaconda3\envs\tok\Lib\site-packages\kedro\ipython\__init__.py\__init__.py]8;;\:]8;id=813320;file://c:\Users\bcech\anaconda3\envs\tok\Lib\site-packages\kedro\ipython\__init__.py#119\119]8;;\

                    INFO     Defined global variable 'context', 'session', 'catalog' and            ]8;id=499279;file://c:\Users\bcech\anaconda3\envs\tok\Lib\site-packages\kedro\ipython\__init__.py\__init__.py]8;;\:]8;id=324617;file://c:\Users\bcech\anaconda3\envs\tok\Lib\site-packages\kedro\ipython\__init__.py#120\120]8;;\
                             'pipelines'                                                                           

                    INFO     Registered line magic 'run_viz'                                        ]8;id=98519;file://c:\Users\bcech\anaconda3\envs\tok\Lib\site-packages\kedro\ipython\__init__.py\__init__.py]8;;\:]8;id=893760;file://c:\Users\bcech\anaconda3\envs\tok\Lib\site-packages\kedro\ipython\__init__.py#126\126]8;;\

In [2]:
shakespeare = catalog.load("shakespeare")

[02/06/24 23:36:50] INFO     Loading data from shakespeare (TextDataset)...                     ]8;id=513492;file://c:\Users\bcech\anaconda3\envs\tok\Lib\site-packages\kedro\io\data_catalog.py\data_catalog.py]8;;\:]8;id=283841;file://c:\Users\bcech\anaconda3\envs\tok\Lib\site-packages\kedro\io\data_catalog.py#483\483]8;;\

In [3]:
shakespeare[:100]

'First Citizen:\nBefore we proceed any further, hear me speak.\n\nAll:\nSpeak, speak.\n\nFirst Citizen:\nYou'

# split on words and specific punctuation

In [4]:
delimiters = r";|,|\n|'|`| "
pat = re.compile(delimiters)

s = shakespeare[:100]

In [5]:
s

'First Citizen:\nBefore we proceed any further, hear me speak.\n\nAll:\nSpeak, speak.\n\nFirst Citizen:\nYou'

In [6]:
t = re.split(pat, s[:100])
print(t)

['First', 'Citizen:', 'Before', 'we', 'proceed', 'any', 'further', '', 'hear', 'me', 'speak.', '', 'All:', 'Speak', '', 'speak.', '', 'First', 'Citizen:', 'You']


Though we don't have any contractions, we can see words are roughly whole. The next step, would be to assign a unique integer to each word. 
For us, a good-enough is the entire dataset vocabulary.

In [7]:
count_vec = CountVectorizer(analyzer='word', tokenizer=lambda x: re.split(pat, x), lowercase=False, stop_words=None)
count_vec.fit([shakespeare])
# print the vocabulary encoding values, sorted
print(f"{len(count_vec.vocabulary_) = }")
{k: v for k, v in sorted(count_vec.vocabulary_.items(), key=lambda item: item[1])}

                    WARNING  c:\Users\bcech\anaconda3\envs\tok\Lib\site-packages\sklearn\feature_ex ]8;id=945673;file://c:\Users\bcech\anaconda3\envs\tok\Lib\warnings.py\warnings.py]8;;\:]8;id=135512;file://c:\Users\bcech\anaconda3\envs\tok\Lib\warnings.py#109\109]8;;\
                             traction\text.py:525: UserWarning: The parameter 'token_pattern' will                 
                             not be used since 'tokenizer' is not None'                                            
                               warnings.warn(                                                                      
                                                                                                                   

len(count_vec.vocabulary_) = 19467


{'': 0,
 '!': 1,
 '&C:': 2,
 '&c.': 3,
 '--': 4,
 '--Ding-dong': 5,
 '--Hold': 6,
 '--I': 7,
 '--O': 8,
 '--Plague': 9,
 '--Tybalt': 10,
 '--Where': 11,
 '--a': 12,
 '--an': 13,
 '--and': 14,
 '--as': 15,
 '--be': 16,
 '--believe': 17,
 '--but': 18,
 '--cast': 19,
 '--cousin': 20,
 '--do': 21,
 '--fellow': 22,
 '--for': 23,
 '--give': 24,
 '--goddess!--O': 25,
 '--here': 26,
 '--how': 27,
 '--if': 28,
 '--it': 29,
 '--let': 30,
 '--like': 31,
 '--mightst': 32,
 '--or': 33,
 '--part': 34,
 '--poor': 35,
 '--so': 36,
 '--that': 37,
 '--the': 38,
 '--then': 39,
 '--there': 40,
 '--thine': 41,
 '--this': 42,
 '--thou': 43,
 '--thus': 44,
 '--we': 45,
 '--what': 46,
 '--whate': 47,
 '--which': 48,
 '--whither': 49,
 '--whom': 50,
 '--will': 51,
 '--you': 52,
 '--your': 53,
 '3': 54,
 ':': 55,
 '?': 56,
 'A': 57,
 'ABHORSON:': 58,
 'ABRAHAM:': 59,
 'ADRIAN:': 60,
 'AEacides': 61,
 'AEdile:': 62,
 'AEdiles': 63,
 'AEdiles!': 64,
 'AEneas': 65,
 'AEsop': 66,
 'ALL': 67,
 'ALL:': 68,
 'ALONSO:'

In [8]:
count_vec.vocabulary_['plague']

13669

In [9]:
def encoding(s: List[str], cv) -> List[int]:
    """s is the pre-split string. """
    vocab = cv.vocabulary_
    encoded = [vocab[tok] for tok in s if tok in vocab]
    return encoded

In [10]:
my_encoding = encoding(t, count_vec)
print(my_encoding)

[1151, 653, 361, 18755, 14061, 3999, 9160, 0, 9944, 11998, 16235, 0, 172, 2816, 0, 16235, 0, 1151, 653, 3436]


Note the `0`s, those are the spaces. 

## Thoughts

The idea is interesting, each word hold a _lot_ of contextual information. That is, each word, even a compound word, feel "unique" to the space. In the case of the Shakespeare dataset, this is `N=19467` unique tokens given our delimiter splitting. A real English unigram vocabulary could be hundreds of thousands of tokens in cardinality. 

Limits: lack of shared meaning across similar words. `count_vec.vocabulary_['plagues']=13672` while `plague=13669`. (numeric distance does not represent similarity)

### Stopwords

We may want to keep vocabulary smaller. We can drop "stop" words. words that connect or are frequently used. Stop words are words like “and”, “the”, “him”, which are presumed to be uninformative in representing the content of a text, and which may be removed to avoid them being construed as signal for prediction.

In [11]:
print(stopwords.words('english'))

['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd", 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', "she's", 'her', 'hers', 'herself', 'it', "it's", 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', "that'll", 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'off', 'over', 'under', 'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why', 'how', 'all', 'any', 'both', 'each', 'few', 'more', 'most', 'other', 'some', 'such', 'no', 'nor', 'not', 'only', 'own', 'same', 'so', 'than', '